_posts/2024-10-24-why-new-rtos.markdown

```diff
@@ -0,0 +1,217 @@
+Unfortunately, as we looked at ThreadX we found that there was a lot of tight coupling that made compartmentalisation difficult.
+If we did it, the result would look so unlike ThreadX that we wouldn't get much benefit.
+It would not be compatible with existing ThreadX uses.
+
+We also considered some other potential options but found similar issues.
+
+## Security is not an add-on
+
+We believe that building a secure system means enabling and respecting two key principles:
+
+ - *The principle of least privilege* means that no component should run with privileges that it doesn't need.
+ - *The principle of intentional use* means that no component should accidentally exercise privileges that it needs for some purpose other than its current operation.
+
+CHERI is designed around these principles and CHERIoT inherits them.
+The principle of least privilege is embodied in CHERI by allowing you to hand a pointer that grants access to a single byte of memory, perhaps with only read permission, if that's all that a function requires.
+The principle of intentional use is embodied by CHERI because you must use the *correct* capability as the operand to load or store instructions.
+Just because you hold pointers to two objects doesn't mean that an out-of-bounds access to one will let you access the other.
+In contrast, memory management unit (MMU) or memory protection unit (MPU) designs may enable least privilege (though not in a way that's convenient for programmers: more on this later) but they can't enforce the principle of intentional use.
+If I have access to two memory regions protected by an MPU and I do some invalid pointer arithmetic from one, I can access the other.
+
+CHERIoT extends these principles into the software space.
+I'll talk later about how we build a software capability model to authorise operations at higher levels of abstraction than 'can I read or write this object' but the principle of least privilege permeates the design.
+
+As with microkernel designs, we do not have a single large privileged entity.
+Unlike most microkernel designs, we don't even have a single *small* entity in the system that holds the rights do absolutely anything (at least, not after the bootloader has run, and the bootloader is deterministic and does not process any data except that in the firmware image).
+
+The most privileged entity in the system is the *switcher*.
+This is a few hundred instructions and we're starting to work with some folks interested in formally verifying it (let us know if you want to join in this activity!).
+The main privilege that this holds is the ability to access the register that contains a capability to the register-save area for context switching threads and the trusted stack used to enforce call-return discipline on cross-compartment calls.
```

Comment:
Maybe a one-liner about the distinction between a thread and compartment would be useful to understand the implications of this statement (without having to read the docs)